Add Docker support

aperture147 opened the pull request:
Things has been made so far:

- A Dockerfile based on the latest LTS `ubuntu:focal`
- A Dockerfile based on older LTS `ubuntu:bionic`
- A simple `docker-compose.yml` file
- Modified README.md, `docker-compose` deployment example added

aperture147 committed:
Add docker related files, Edit README.md

aperture147 committed:
Update dockerfile, add options to specity the git repo and git branch, also update README to suit the changes

aperture147 committed:
Fix readmy.md grammar

aperture147 committed:
change current dockerfile base image to ubuntu:bionic
Add new dockerfile using ubuntu:focal

aperture147 committed:
add mysql to docker-compose file

aperture147 committed:
Merge remote-tracking branch 'upstream/develop' into develop

aperture147 committed:
Change workdir to /usr/local/sbin

aperture147 committed:
Add bnetd, d2cs and d2dbs to docker-compose
Update README.md, add docker-compose example

aperture147 committed:
default image now use focal

aperture147 commented:
@RElesgoe Please check this out.

aperture147 committed:
Switch to quay.io/aperture147/ubuntu to mitigate the impact of Docker Hub Pull Request Limits
Add some git options to decrease the clone time
fix git clone directory to pvpgn-server

aperture147 commented:
Some changes were made:
- Switch base image from default `ubuntu` to `quay.io/aperture147/ubuntu` to [mitigate the impact of Docker Hub Pull Request Limits](https://docs.docker.com/docker-hub/download-rate-limit/) which can make CI build failed. This is an automated build from a synced-fork [aperture147/docker-brew-ubuntu-core](https://github.com/aperture147/docker-brew-ubuntu-core).
- Fix a bug which the cloned repo could be placed in a different directory
- Decrease build time by adding `--depth=1` to the `git clone` command

aperture147 committed:
Revert -H option of cmake

aperture147 committed:
Add more manual for docker deployment to readme.md

RElesgoe commented:
When running one of the example commands,

```
docker build . --build-arg with_d2cs=true --build-arg with_bnetd=false --build-arg with_mysql=false --build-arg with_sqlite3=true --build-arg git_branch=develop -t pvpgn-server:d2cs-sqlite
```

The build errors with:

```
Cloning into 'pvpgn-server'...
error: pathspec 'develop' did not match any file(s) known to git
The command '/bin/sh -c apt-get update &&     apt-get install -y --no-install-recommends         git         netbase         ca-certificates         gcc g++         libc6-dev libc6         libc++-dev libc++1          zlib1g-dev zlib1g         libcurl4-openssl-dev libcurl4         cmake make         $(if ${with_mysql}; then echo "libmysqlclient-dev libmysqlclient21"; fi)         $(if ${with_sqlite3}; then echo "libsqlite3-dev libsqlite3-0"; fi)         $(if ${with_pgsql}; then echo "libpq-dev libpq5"; fi)         $(if ${with_odbc}; then echo "unixodbc-dev libodbc1"; fi)         $(if ${with_lua}; then echo "liblua5.1-0-dev liblua5.1-0"; fi) &&     git clone --depth=1 ${git_repo} pvpgn-server &&     cd pvpgn-server &&     git checkout ${git_branch} &&     cmake -G "Unix Makefiles" -S./ -B./build           -DWITH_BNETD=${with_bnetd}           -DWITH_D2CS=${with_d2cs}           -DWITH_D2DBS=${with_d2dbs}           -DWITH_LUA=${with_lua}           -DWITH_MYSQL=${with_mysql}           -DWITH_SQLITE3=${with_sqlite3}           -DWITH_PGSQL=${with_pgsql}           -DWITH_ODBC=${with_odbc} &&     cd build && make -j$(nproc) && make install &&     apt-get autoremove --purge -y         git         gcc g++         libc6-dev         libc++-dev         zlib1g-dev         libcurl4-openssl-dev         cmake make         $(if ${with_mysql}; then echo "libmysqlclient-dev "; fi)         $(if ${with_sqlite3}; then echo "libsqlite3-dev "; fi)         $(if ${with_pgsql}; then echo "libpq-dev "; fi)         $(if ${with_pgsql}; then echo "unixodbc-dev"; fi)         $(if ${with_lua}; then echo "liblua5.1-0-dev"; fi) &&     rm -rf /var/lib/apt/lists/* /build/pvpgn-server &&     apt-get clean' returned a non-zero code: 1
```

aperture147 commented:
> 
> 
> When running one of the example commands,
> 
> ```
> docker build . --build-arg with_d2cs=true --build-arg with_bnetd=false --build-arg with_mysql=false --build-arg with_sqlite3=true --build-arg git_branch=develop -t pvpgn-server:d2cs-sqlite
> ```
> 
> The build errors with:
> 
> ```
> Cloning into 'pvpgn-server'...
> error: pathspec 'develop' did not match any file(s) known to git
> The command '/bin/sh -c apt-get update &&     apt-get install -y --no-install-recommends         git         netbase         ca-certificates         gcc g++         libc6-dev libc6         libc++-dev libc++1          zlib1g-dev zlib1g         libcurl4-openssl-dev libcurl4         cmake make         $(if ${with_mysql}; then echo "libmysqlclient-dev libmysqlclient21"; fi)         $(if ${with_sqlite3}; then echo "libsqlite3-dev libsqlite3-0"; fi)         $(if ${with_pgsql}; then echo "libpq-dev libpq5"; fi)         $(if ${with_odbc}; then echo "unixodbc-dev libodbc1"; fi)         $(if ${with_lua}; then echo "liblua5.1-0-dev liblua5.1-0"; fi) &&     git clone --depth=1 ${git_repo} pvpgn-server &&     cd pvpgn-server &&     git checkout ${git_branch} &&     cmake -G "Unix Makefiles" -S./ -B./build           -DWITH_BNETD=${with_bnetd}           -DWITH_D2CS=${with_d2cs}           -DWITH_D2DBS=${with_d2dbs}           -DWITH_LUA=${with_lua}           -DWITH_MYSQL=${with_mysql}           -DWITH_SQLITE3=${with_sqlite3}           -DWITH_PGSQL=${with_pgsql}           -DWITH_ODBC=${with_odbc} &&     cd build && make -j$(nproc) && make install &&     apt-get autoremove --purge -y         git         gcc g++         libc6-dev         libc++-dev         zlib1g-dev         libcurl4-openssl-dev         cmake make         $(if ${with_mysql}; then echo "libmysqlclient-dev "; fi)         $(if ${with_sqlite3}; then echo "libsqlite3-dev "; fi)         $(if ${with_pgsql}; then echo "libpq-dev "; fi)         $(if ${with_pgsql}; then echo "unixodbc-dev"; fi)         $(if ${with_lua}; then echo "liblua5.1-0-dev"; fi) &&     rm -rf /var/lib/apt/lists/* /build/pvpgn-server &&     apt-get clean' returned a non-zero code: 1
> ```

Oh sorry, I forgot to commit some changes. Btw I'm on a Lunar New Year vacation and don't bring any computer with. I will fix this problem as soon as possible

aperture147 committed:
Fix git branch checking out
will not work with depth=1

aperture147 commented:
@RElesgoe the problem is now fixed.

aperture147 committed:
Correct some explanation comments

Happin3ss commented on Dockerfile:
```diff
@@ -0,0 +1,76 @@
+FROM quay.io/aperture147/ubuntu:focal
```
@aperture147 Can we use container from docker hub instead of personal container ?

trungklam commented on Dockerfile:
```diff
@@ -0,0 +1,76 @@
+ARG with_sqlite3=false
+ARG with_pgsql=false
+ARG with_odbc=false
+ARG with_lua=false
```
Should lua support enable as default ?

aperture147 replied:
~~In my build, enable LUA will cause pvpgn server to crash everytime an user connect~~ Not anymore if you use `develop` branch. The crash is not related to Lua since it happens in any build configuration if you build on branch `master`
Btw LUA is a experimental feature and an average server admin will not gain any benefit from that so I don't think we should enable this by default.

aperture147 replied:
> Should lua support enable as default ?

Sorry, for the false claim, server crashing is not related to Lua since it happens on any build configuration if you build this image on branch `master`. I've just created [a issue](https://github.com/pvpgn/pvpgn-server/issues/446) for this

RElesgoe commented:
> Switch base image from default ubuntu to quay.io/aperture147/ubuntu to mitigate the impact of Docker Hub Pull Request Limits which can leads to failing CI build. This is an automated build from a synced-fork aperture147/docker-brew-ubuntu-core.

I prefer using the official Ubuntu image from the default container repository. If a user is experiencing rate limiting, they should modify the Dockerfile on their machine to avoid rate limiting.

After running
```
docker run -v /tmp/conf:/tmp/conf --rm --entrypoint cp pvpgn-server:bnetd-mysql /usr/local/etc/pvpgn/* /tmp/conf
```
I receive the following error:
```
cp: -r not specified; omitting directory '/usr/local/etc/pvpgn/i18n'
```

I receive similar errors when running
```
docker run -v /tmp/assets:/tmp/assets --rm --entrypoint cp pvpgn-server:bnetd-mysql /usr/local/var/pvpgn/* /tmp/assets
```

I also noticed that one of the commands specifies the `mysql:8.0.23` image. Can we use `mysql:8.0` or `mysql:8` instead?


After running
```
docker run -d --name pvpgn-d2cs --restart unless-stopped -p 6113:6113 -p 6113:6113/udp -v /tmp/conf:/usr/local/etc/pvpgn -v /tmp/assets:/usr/local/var/pvpgn pvpgn-server:d2cs-mysql
```
I receive the following error:
```
docker: Error response from daemon: OCI runtime create failed: container_linux.go:370: starting container process caused: exec: "bnetd": executable file not found in $PATH: unknown.
```

When checking the logs using
```
docker logs -f --tail 1000 pvpgn-bnetd
```
All I get is:
```
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
You are currently running PvPGN 1.99.7.2.1-PRO
=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
If you need support:
 * Create an issue at https://github.com/pvpgn/pvpgn-server

Server is now running.
```
Because `bnetd` is running in the foreground.

aperture147 committed:
Switch back to official ubuntu image

aperture147 committed:
Fix copy/paste errors

aperture147 committed:
Fix to mysql:8.0

aperture147 committed:
Fix copy command mistake

aperture147 committed:
Add SERVER_TYPE environment to switch between bnetd, d2cs and d2dbs

aperture147 committed:
Update README.md

aperture147 commented:
### About changing the base image to official `ubuntu`:
Changed. According to [this document](https://docs.docker.com/docker-hub/download-rate-limit/#how-can-i-check-my-current-rate), the rate limit is 100 pull per 6 hours so this might not affect this project's CI/CD

### About getting error while copying base config files from the image:
Fixed. I forgot to add `-r`

### About changing `mysql` version to `8.0` or `8`:
In my opinion, locking `mysql` (or any image) to a specific version is a best practice, this might help us to figure the `best-known version` and make the debugging process easier. Changing `mysql` to `latest` or `8` will pull the latest version (currently 8.0 but could be 8.1 or 8.2 in the future) with (maybe) some breaking changes. Although `mysql` is pretty stable and `libmysqlclient` supports old version pretty well but to keep the best practice, I will only change the `mysql` version tag to `8.0`.

### About running specific build of `d2cs` and `d2dbs` cause error due to `bnetd` missing:
This is my fault to hard-coded the CMD in the Dockerfile to `bnetd -f`. It's hard to figure out which server type the operator want to build, they can build an image which supports any combination of `bnetd`, `d2cs` and `d2dbs` so writing a script to predict the CMD command is not possible. My solution is adding another `SERVER_TYPE` env var to the image (which is set default to `bnetd`) and modify the CMD to `$SERVER_TYPE`. Anytime an operator wants to spin up a server, they have to choose an appropriate server type between between `bnetd`, `d2cs` and `d2dbs`. I already push this change, do you have any better solution?

aperture147 committed:
Add more comment

aperture147 commented:
Another note, this command:

```bash
docker logs -f --tail 1000 pvpgn-bnetd
```

will continuously fetch the command with 1000 last lines of log of **pvpgn-bnetd** container, not the **pvpgn-d2cs** or **pvpgn-d2dbs* container. The error you get while running **pvpgn-d2cs** or **pvpgn-d2dbs** happens because the build doesn't include **bnetd**, which is mentioned in the "About running specific build of d2cs and d2dbs cause error due to bnetd is missing" part of the previous comment.

aperture147 committed:
Separate Docker readme to another file

aperture147 committed:
Make warning more noticeable

aperture147 committed:
Fix text coloring

aperture147 committed:
Remove unrelated comment

aperture147 committed:
fix some minor changes on readme_docker.md

aperture147 commented:
Do you guys use anything like discord, whatsapp or even IRC so we can contact more convenient.

RElesgoe commented:
> Do you guys use anything like discord, whatsapp or even IRC so we can contact more convenient.

**#botdev** at [Discord - BNETDocs](https://discord.com/invite/u87WVeu)

aperture147 committed:
use -D to send all log to stdout

aperture147 committed:
Switch branch to develop

aperture147 committed:
Fix copy command

cen1 commented on README_DOCKER.md:
```diff
@@ -0,0 +1,223 @@
+Player vs Player Gaming Network - PRO on Docker
+=====
+
+To quickly deploy a PVPGN server on linux, using Docker is one of the best choices. You can gain some benefits using this:
```
Remove linux keyword since docker is not linux specific. We can run this on windows and mac os I assume (testing).

cen1 commented on README_DOCKER.md:
```diff
@@ -0,0 +1,223 @@
+- Quick deploy, easy to scale out, manage and update
+- Simple to use and config
+
+## Prequisite:
```
spelling

aperture147 replied:
Sorry I'm not a native english speaker. I don't know which part is having a spelling error :(

RElesgoe replied:
Prequisite should be Prerequisite

cen1 replied:
@aperture147 what relesgoe said

cen1 commented on README_DOCKER.md:
```diff
@@ -0,0 +1,223 @@
+# create configuration and asset locations
+mkdir -p /your/config/dir /your/assets/dir
+# copy configuration files and base assets from an image, all base configuration files in any image are the same
+docker run -v /your/config/dir:/tmp/conf --rm --entrypoint cp pvpgn-server:bnetd-mysql -r /usr/local/etc/pvpgn/* /tmp/conf
```
Pretty sure the image name should be at the end otherwise it is grabbed by cp command. I get:

sudo docker run -v /opt/pvpgn/etc:/tmp/conf --rm --entrypoint cp pvpgn-server:bnetd-mysql -r /usr/local/etc/pvpgn/* /tmp/conf
cp: cannot stat '/usr/local/etc/pvpgn/bnetd_default_user.cdb': No such file or directory

aperture147 replied:
> 
> 
> Pretty sure the image name should be at the end otherwise it is grabbed by cp command. I get:
> 
> sudo docker run -v /opt/pvpgn/etc:/tmp/conf --rm --entrypoint cp pvpgn-server:bnetd-mysql -r /usr/local/etc/pvpgn/* /tmp/conf
> cp: cannot stat '/usr/local/etc/pvpgn/bnetd_default_user.cdb': No such file or directory

I've already overwrite the entrypoint of the container to `cp`. Normally it will be `/bin/sh`. If you run a docker command like this:
```sh
docker run some-image some-command
```
It means that you are running `/bin/sh some-command`. But changing the entrypoint to `cp`, running a command like this:
```sh
docker run --entrypoint cp some-image -r /dir1/* /dir2
```
the container will execute your command like this: `cp -r /dir1/* /dir2`. That's why you can run `mysql` image like this:
```sh
docker run mysql --default-authentication-plugin=mysql_native_password
```

Actually the error is caused by `cp` command somehow couldn't recognize the wildcard directory and got it as a directory or file. I don't know why this problem happens. Fixing it is pretty easy, just change the `-r /usr/local/etc/pvpgn/* /tmp/conf` to `-a /usr/local/etc/pvpgn/. /tmp/conf`.

cen1 commented on docker-compose.yml:
```diff
@@ -0,0 +1,50 @@
+version: "3"
+
+services:
+  bnetd:
```
Missing `depends_on` so mysql always starts first.

cen1 commented on docker-compose.yml:
```diff
@@ -0,0 +1,50 @@
+      - 6112:6112/udp
+      - 6112:6112
+    volumes:
+      - /your/config/dir:/usr/local/etc/pvpgn
```
How about we change this to something more probable, let's say `/opt/pvpgn/etc` and `/opt/pvpgn/var`?

cen1 commented:
If I run with `docker run pvpgn-server:bnetd-mysql` I cannot kill the container with ctrl+c. Something catches the signal?

cen1 reviewed (commented):
If I run with docker run pvpgn-server:bnetd-mysql I cannot kill the container with ctrl+c. Something catches the signal?

cen1 commented:
Dockerfile is good, small fix to compose and mostly readme fixes which I documented above. After next iteration I will also retest on Mac and Windows since it should work just fine in theory.

Best regards.

aperture147 commented:
> 
> 
> If I run with docker run pvpgn-server:bnetd-mysql I cannot kill the container with ctrl+c. Something catches the signal?

I'm running it as `bnetd -D`, so it's the one which catching the stop signal should be the `bnetd` itself. The default signal which docker sends is `SIGTERM`, after a few seconds or you hit the ctrl + c again then it will send `SIGKILL`. I don't know what is the meaning of "killing" the container, is it gracefully stop the container or make it stop immediately?

aperture147 committed:
Merge remote-tracking branch 'upstream/develop' into develop

aperture147 committed:
Change config location to /opt/pvpgn

aperture147 committed:
Update readme_docker, change placeholder directory in commands to /opt/var/*, also fix configuration obtaining command.

aperture147 commented:
> 
> 
> Dockerfile is good, small fix to compose and mostly readme fixes which I documented above. After next iteration I will also retest on Mac and Windows since it should work just fine in theory.
> 
> Best regards.

Thanks! I've resolved all of the issues mentioned. This build is tested on Linux and I'm running it on my server which has approx 200 - 300 unique users daily (most of them play DotA with my hosted ghost++).

If this build works on linux then 99% it will run on windows and macos too since docker on those system is running on a VM. If there are errors, most of the time it would be networking misconfiguration.

cen1 commented on Dockerfile:
```diff
@@ -0,0 +1,77 @@
+       6114/udp
+
+ENV SERVER_TYPE bnetd
+CMD $SERVER_TYPE -D
```
Replace with `CMD exec $SERVER_TYPE -D`

This replaces sh with bnetd as main process. It also makes ctrl+c work as expected. Shouldn't have any negative consequences. See:
https://github.com/moby/moby/issues/2838
https://askubuntu.com/questions/525767/what-does-an-exec-command-do

cen1 commented on bionic.dockerfile:
```diff
@@ -0,0 +1,78 @@
+       6114/udp
+
+ENV SERVER_TYPE bnetd
+CMD $SERVER_TYPE -D
```
Add exec, same as Dockerfile comment

cen1 commented on README_DOCKER.md:
````diff
@@ -0,0 +1,207 @@
+```
+Modifying the copied configuration files to fit your needs. check [this link](https://pvpgn.pro/pvpgn_installation.html) to get more information of the configuration files.
+
+### Using docker only:
````
I think we are missing a crucial piece of information here: how to modify bnetd (and others?) configuration so it connects to mysql docker container. I tried: `storage_path = "sql:mode=mysql;host=mysql;name=pvpgn-bnetd;user=bnetd;pass=example;default=0;prefix=pvpgn_"` but it can't connect. What config did you use?

aperture147 replied:
have you tried the the `docker-compose.yml`? I think it should run in one click. I will double check this.

cen1 replied:
Yes, unless I can properly set the config to use mysql it just uses file based storage.

cen1 commented:
Left a couple more comments. Just need to test Windows and we are almost there.

aperture147 commented:
Finished!

cen1 commented:
@aperture147 please check the 4 pending comments above. The mysql config string is the important one, the rest I can close myself.

aperture147 commented:
> @aperture147 please check the 4 pending comments above. The mysql config string is the important one, the rest I can close myself.

But I don't see any new comment :( Resolved all 5 issues above. Am I missing anything?
![image](https://user-images.githubusercontent.com/32543264/112140011-0b2b1f80-8c06-11eb-919b-646e3ab28a6c.png)

cen1 commented:
> > @aperture147 please check the 4 pending comments above. The mysql config string is the important one, the rest I can close myself.
> 
> But I don't see any new comment :( Resolved all 5 issues above. Am I missing anything?
> ![image](https://user-images.githubusercontent.com/32543264/112140011-0b2b1f80-8c06-11eb-919b-646e3ab28a6c.png)

My bad, I forgot to actually submit the review. :insert facepalm:

aperture147 committed:
resolve conversations

aperture147 commented:
just resolved your issue above. Please check!

cen1 commented:
I had some issues with modern firewalld blocking internal docker container comms but I will patch the readme mysql with that info and fix the language a little bit. All seems to work fine, I made it to the login.
